### PR TITLE
HW

### PR DESCRIPTION
**Pull request by Chorniypistolet**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -33,18 +33,24 @@
 			<version>8.0.33</version>
 		</dependency>
 		<dependency>
-			<groupId>org.springframework.boot</groupId>
-			<artifactId>spring-boot-starter-data-jpa</artifactId>
+			<groupId>org.liquibase</groupId>
+			<artifactId>liquibase-core</artifactId>
+			<version>4.29.0</version>
```
Comment (Elena-Bruyako) [inline pom.xml]:
```suggestion
			<version>${liquibase.version}</version>
```

spring boot parent provides this version

File: pom.xml [hunk 2]
```diff
@@ -74,6 +80,11 @@
 			<artifactId>springfox-core</artifactId>
 			<version>3.0.0</version>
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-security</artifactId>
+			<version>3.3.2</version>
```
Comment (Elena-Bruyako) [inline pom.xml]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/**", "/error")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/config/SecurityConfig.java]:
```suggestion
                                .requestMatchers("/auth/**", "/error", "/swagger-ui/**", "/v3/api-docs/**")
```

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 4]
```diff
@@ -6,27 +6,22 @@
...
-import org.springframework.web.bind.annotation.ResponseStatus;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
 
 @Tag(name = "Book API")
 @RequiredArgsConstructor
 @RestController
-@RequestMapping("/books")
+@RequestMapping("/api/books")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]:
```suggestion
@RequestMapping("/books")
```

just add `server.servlet.context-path=/api` to application.properties

File: src/main/java/mate/academy/spring/boot/exception/BasicCredentialException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.spring.boot.exception;
+
+public class BasicCredentialException extends RuntimeException{
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/exception/BasicCredentialException.java]: Where do you use it?

File: src/main/java/mate/academy/spring/boot/model/User.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    private String shippingAddress;
+    @Column(name = "is_deleted", columnDefinition = "TINYINT")
+    private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.EAGER)
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/mate/academy/spring/boot/model/User.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    @Column(nullable = false)
+    private String password;
+    @Column(nullable = false)
+    private String firstName;
+    @Column(nullable = false)
+    private String lastName;
+    private String shippingAddress;
+    @Column(name = "is_deleted", columnDefinition = "TINYINT")
+    private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/spring/boot/model/User.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return List.of();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/User.java]:
```suggestion
        return roles;
```

File: src/main/java/mate/academy/spring/boot/model/Role.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+
+@Entity
+@Getter
+@Setter
+@ToString
+@Table(name = "roles")
+public class Role {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Role.java]:
```suggestion
public class Role implements GrantedAuthority {
```

File: src/main/java/mate/academy/spring/boot/security/CustomUserDetailsService.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(username)
+                .orElseThrow(() -> new UsernameNotFoundException("Can't find user by username: " + username));
+        Set<GrantedAuthority> authorities = user.getRoles().stream()
+                .map(role -> new SimpleGrantedAuthority(role.getName().name()))
+                .collect(Collectors.toSet());
+        return new org.springframework.security.core.userdetails.User(user.getEmail(), user.getPassword(), authorities);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/security/CustomUserDetailsService.java]:
```suggestion
        return repository.findByEmail(username)
                .orElseThrow(() -> new EntityNotFoundException(
                        "Can't find user by email: " + username));
```

File: src/main/java/mate/academy/spring/boot/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) throws RegistrationException {
+        if(userRepository.existsByEmail(requestDto.getEmail())) {
+            throw new RegistrationException(String.format("User with this email: %s already exists"
+                    , requestDto.getEmail()));
+        }
+        User user = userMapper.toUser(requestDto);
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
+        User savedUser = userRepository.save(user);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
```

File: src/main/java/mate/academy/spring/boot/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        return userMapper.toUserResponse(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/UserServiceImpl.java]:
```suggestion
        return userMapper.toUserResponse(user);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/spring/boot/controller/AuthenticationController.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/auth")
+@RequiredArgsConstructor
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    public UserResponseDto register(@RequestBody UserRegistrationRequestDto requestDto) throws RegistrationException {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/AuthenticationController.java]:
```suggestion
    public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto requestDto) throws RegistrationException {
```

File: src/main/java/mate/academy/spring/boot/model/Role.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false, unique = true)
+    private RoleName name;
+
+    @Override
+    public String getAuthority() {
+        return null;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Role.java]: why is null?

File: src/main/java/mate/academy/spring/boot/security/CustomUserDetailsService.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByEmailWithRoles(username)
+                .orElseThrow(() -> new UsernameNotFoundException("Can't find user by username: " + username));
+        Set<GrantedAuthority> authorities = user.getRoles().stream()
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/security/CustomUserDetailsService.java]: see previous comment

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments
You have conflicts in your branch, please fix them

File: src/main/java/mate/academy/spring/boot/security/CustomUserDetailsService.java (hunk 15 above)
Comment (Chorniypistolet) [reply src/main/java/mate/academy/spring/boot/security/CustomUserDetailsService.java]: but if I don't use this variant, then I have problems with endpoints that are available only to the admin

File: src/main/resources/db/changelog/changes/08-update-users-table.yaml [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+        - sql:
+            dbms: mysql
+            endDelimiter: ";"
+            splitStatements: true
+            stripComments: true
+            sql: |
+              INSERT INTO users_roles (user_id, role_id)
+              SELECT id, 1
+              FROM users
+              WHERE id NOT IN (SELECT user_id FROM users_roles)
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/08-update-users-table.yaml]:
```suggestion
            sql: |
              INSERT INTO users_roles (user_id, role_id)
              SELECT id, 1
              FROM users
              WHERE id NOT IN (SELECT user_id FROM users_roles)

```
image
go through the files in your PR, add empty line where needed

File: src/main/resources/db/changelog/changes/06-add-admin-users-table.yaml [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+      changes:
+        - insert:
+            tableName: users
+            columns:
+              - column:
+                  name: email
+                  value: "admin@mail.com"
+              - column:
+                  name: password
+                  value: "$2a$10$Y4q3wbO7NeW8eWRaE0jqMuIKIlmMNC1qt6ypQdTNPtgKq00rk3.Ki"
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/06-add-admin-users-table.yaml]: add password in plantext as comment, it's useless otherwise

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/spring/boot/service/impl/BookServiceImpl.java [hunk 18]
```diff
@@ -29,6 +32,7 @@ public BookDto save(CreateBookRequestDto requestDto) {
     }
 
     @Override
+    @Transactional(readOnly = true)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/BookServiceImpl.java:35]: redundant

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**